### PR TITLE
Upgrade to `rand` 0.8 everywhere

### PR DESCRIPTION
**Pull request by Twey**
## Motivation

We previously had an old version of `rand` pinned in several crates as `rand07`.  This version had several disadvantages, including being more error-prone when used in conditional builds (it unconditionally exported features that would panic at runtime).

## Proposal

Upgrade `rand` to 0.8 across the board.  In order to do this we:

- bump `ed25519` to 2.2
- bump `ed25519-dalek` to 2.1.1
- bump `ahash` to 0.7.8 and also 0.8.8 (we use both 0.7 and 0.8 transitively)
- bump `crc32c` to 0.6.5
- bump the nightly Rust to 2024-02-14

One might reasonably express surprise, especially at the last update.  Because we are on the nightly treadmill for our lint workflow, we get no backwards compatibility guarantees with regards to stable features.  Since a number of our dependencies have just jumped a compatibility gap with respect to the `stdsimd` feature, we must upgrade the compiler and all packages that depend on that feature simultaneously — we have neither forwards nor backwards compatibility.  Bumping them all in one PR is the only way to avoid breaking CI.

Some dead code is also eliminated, as Clippy's dead code detection gets more sophisticated.  The `linera_witty::tests::common::test_instance::WithoutExports` struct is marked `#[allow(dead_code)]`, as otherwise Clippy complains about its constructor being unused.

## Test Plan

CI.

## Release Plan

Should be invisible to users.

## Links

- [reviewer checklist](https://github.com/linera-io/linera-protocol/blob/main/CONTRIBUTING.md#reviewer-checklist)

### CODE REVIEW

**Review by jvff — APPROVED** [review]
Nice work! :tada:

**Comment by Twey** [comment]
The `crowd-funding` README had to be updated because the `StdRng` RNG algorithm changed between `rand` 0.7 and 0.8 (still ChaCha, but from 20 rounds to 12).  If we want to avoid the READMEs breaking in future version upgrades, we should use `rand_chacha` directly, which is guaranteed to be stable.

File: .github/workflows/rust.yml [hunk 1]
```diff
@@ -37,7 +37,7 @@ permissions:
 jobs:
   test:
     runs-on: ubuntu-latest-16-cores
-    timeout-minutes: 45
+    timeout-minutes: 90
```
Comment (ma2bd) [inline .github/workflows/rust.yml:40]: That is a bit sad

File: examples/crowd-funding/README.md [hunk 2]
```diff
@@ -86,7 +86,7 @@ A table will be shown with the chains registered in the wallet and their meta-da
 │ Chain Id                                                         ┆ Latest Block                                                                         │
 ╞══════════════════════════════════════════════════════════════════╪══════════════════════════════════════════════════════════════════════════════════════╡
 │ 1db1936dad0717597a7743a8353c9c0191c14c3a129b258e9743aec2b4f05d03 ┆ Public Key:         84eddaaafce7fb923c3b2494b3d25e54e910490a726ad9b3a2228d3fb18f9874 │
-│                                                                  ┆ Owner:              5d1b22ce0fe6a8bd45aa7d10e753a29bf569a2e39f22a63c462300f7dd41f7da │
+│                                                                  ┆ Owner:              b4f8586041a07323bd4f4ed2d758bf1b9a977eabfd4c00e2f12d08a0899485fd
```
Comment (ma2bd) [inline examples/crowd-funding/README.md:89]: missing `|` at the end?

File: examples/matching-engine/src/contract.rs [hunk 3]
```diff
@@ -35,6 +35,17 @@ enum ModifyAmount {
     Partial(Amount),
 }
 
+/// Transfer operation back to the owners
```
Comment (ma2bd) [inline examples/matching-engine/src/contract.rs:38]: thanks!

File: linera-base/src/crypto.rs [hunk 4]
```diff
@@ -23,7 +23,7 @@ use {
 };
 
 /// A signature key-pair.
-pub struct KeyPair(dalek::Keypair);
+pub struct KeyPair(dalek::SigningKey);
```
Comment (ma2bd) [inline linera-base/src/crypto.rs:26]: Interesting

File: linera-execution/src/util/mod.rs [hunk 5]
```diff
@@ -21,14 +21,6 @@ pub trait UnboundedSenderExt<Request> {
     ) -> Result<oneshot::Receiver<Response>, ExecutionError>
     where
         Response: Send;
-
```
Comment (ma2bd) [inline linera-execution/src/util/mod.rs:24]: This cleanup belongs to a separate PR for proper review/discussion in light of #1416

**Review by ma2bd — APPROVED** [review]

File: .github/workflows/rust.yml (hunk 1 above)
Comment (Twey) [reply .github/workflows/rust.yml:40]: Well typically this time it passed in under 45 😅 I guess it was nearly done last time when it got cancelled.

File: linera-execution/src/util/mod.rs (hunk 5 above)
Comment (Twey) [reply linera-execution/src/util/mod.rs:24]: I'd like to factor it into a separate commit so we can choose to revert it later, but it's going to get squashed away anyway 😅

I guess I can comment it out or `#[allow(dead_code)]` for now to minimize the changeset to be manually reverted later.

File: linera-base/src/crypto.rs (hunk 4 above)
Comment (Twey) [reply linera-base/src/crypto.rs:26]: I wonder if it's generally true for asymmetric signature algorithms that you can extract the verifying key from the signing key.

File: linera-execution/src/util/mod.rs (hunk 5 above)
Comment (Twey) [reply linera-execution/src/util/mod.rs:24]: I added a `TODO` and an `allow(dead_code)` for now.

File: .github/workflows/rust.yml (hunk 1 above)
Comment (Twey) [reply .github/workflows/rust.yml]: Back to 45 now.  The cache should catch it.

File: linera-base/src/crypto.rs (hunk 4 above)
Comment (ma2bd) [reply linera-base/src/crypto.rs:26]: I can't find an example where it wouldn't be the case, but the "extraction" might require arithmetic operations.

Comment (Twey) [reply linera-base/src/crypto.rs:26]: I think it does here. But that's the way `dalek` decided to go: there's no `Keypair` any more (though you can still get the keypair as bytes).
